bpo-34670: Add TLS 1.3 post handshake auth

tiran opened the pull request:
Add SSLContext.post_handshake_auth and
SSLSocket.verify_client_post_handshake for TLS 1.3 post-handshake
authentication.

Signed-off-by: Christian Heimes <christian@python.org>q

<!-- issue-number: [bpo-34670](https://www.bugs.python.org/issue34670) -->
https://bugs.python.org/issue34670
<!-- /issue-number -->

fantix commented on Doc/library/ssl.rst:
```diff
@@ -1929,6 +1939,26 @@ to speed up repeated connections from the same clients.
+   request a TLS client certificate at any time after the handshake.
+
+   When enabled on client-side sockets, the client signals the server that
+   is supports post-handshake authentication.
```
(trivial) typo s/is/it

tiran replied:
Thanks, fixed

1st1 commented on Doc/library/ssl.rst:
```diff
@@ -1314,6 +1314,22 @@ SSL sockets also have the following additional methods and attributes:
+.. method:: SSLSocket.verify_client_post_handshake()
+
+   Requests post-handshake authentication (PHA) from a TLS 1.3 client. PHA
+   can only be initiated for a TLS 1.3 connection from a server-side socket,
```
Does it raise an error if it's not TLS 1.3?

tiran replied:
Yes, it does raise an error if any precondition isn't met (not TLS 1.3, called on client side, PHA not enabled, before or during handshake, ...).

1st1 commented on Lib/ssl.py:
```diff
@@ -777,6 +777,10 @@ def version(self):
 
+    if HAS_TLSv1_3:
+        def verify_client_post_handshake(self):
+            return self._sslobj.verify_client_post_handshake()
```
Do you want to always expose the method but raise `NotImplementedError` if no TLS 1.3 is available?

1st1 commented on Lib/ssl.py:
```diff
@@ -1094,6 +1098,13 @@ def unwrap(self):
+            if self._sslobj:
+                return self._sslobj.verify_client_post_handshake()
+            else:
+                raise ValueError("No SSL wrapper around " + str(self))
```
ditto

tiran committed:
bpo-34670: Add TLS 1.3 post handshake auth

Add SSLContext.post_handshake_auth and
SSLSocket.verify_client_post_handshake for TLS 1.3 post-handshake
authentication.

Signed-off-by: Christian Heimes <christian@python.org>
